notdiamond/toolkit/rag/evaluation.py

```diff
@@ -0,0 +1,238 @@
+from typing import Dict, List, Optional, Sequence, Tuple, Union
+
+import optuna
+import pandas as pd
+from langchain_core.callbacks import Callbacks
+from langchain_core.embeddings import Embeddings as LangchainEmbeddings
+from langchain_core.language_models import BaseLanguageModel as LangchainLLM
+from langchain_core.prompt_values import StringPromptValue
+from ragas import EvaluationDataset, SingleTurnSample
+from ragas._analytics import track_was_completed
+from ragas.cost import TokenUsageParser
+from ragas.embeddings.base import BaseRagasEmbeddings
+from ragas.evaluation import evaluate as ragas_evaluate
+from ragas.llms import BaseRagasLLM
+from ragas.metrics.base import Metric
+from ragas.run_config import RunConfig
+from tqdm import tqdm
+
+from notdiamond.llms.config import LLMConfig
+from notdiamond.toolkit.rag.evaluation_dataset import (
+    RAGEvaluationDataset,
+    RAGSample,
+)
+from notdiamond.toolkit.rag.llms import get_llm
+from notdiamond.toolkit.rag.workflow import BaseNDRagWorkflow
+
+
+def get_eval_dataset(
+    test_queries: pd.DataFrame,
+    workflow: BaseNDRagWorkflow,
+    generation_prompt: str = None,
+    generator_llm: Union[LLMConfig, str] = None,
+):
+    """
+    Create a dataset of RAGSample objects to evaluate the performance of a RAG workflow.
+
+    Args:
+        test_queries: A pandas DataFrame with schema implied by the method below.
+        workflow: BaseNDRagWorkflow subclass created by the user.
+
+    Schema for test_queries can be found below
+    https://docs.ragas.io/en/stable/references/evaluation_schema/#ragas.dataset_schema.SingleTurnSample
+    """
+    samples = []
+    for _, row in test_queries.iterrows():
+        query = row["user_input"]
+        reference = row["reference"]
+        generation_prompt = generation_prompt or row.get("generation_prompt")
+        generator_llm = generator_llm or row.get("generator_llm")
```

Comment:
If these are required by `RAGSample` we should double-check for `None` here and raise a `ValueError` if needed.

Reply:
Done